app/scripts/stores/projectStore.js

```diff
@@ -484,6 +525,76 @@ var ProjectStore = Reflux.createStore({
             error: errMsg,
             loading: false
         });
+    },
+
+    startUpload() {
+        this.trigger({
+            uploading: true
+        })
+    },
+
+    startUploadSuccess(uploadId, details) {
+        this.uploads[uploadId] = details;
+        ProjectActions.updateAndProcessChunks(uploadId, null, null);
+    },
+
+    startUploadError(error) {
+        let errMsg = error && error.message ? "Error: " + error : '';
+        this.trigger({
+            error: errMsg,
+            uploading: false
+        });
+    },
+
+    updateAndProcessChunks(uploadId, chunkNum, status) {
+        if (!uploadId && !this.uploads[uploadId]) {
+            return;
+        }
+        let upload = this.uploads[uploadId];
+        let chunks = this.uploads[uploadId].chunks;
+        // update chunk
+        if (chunkNum !== null) {
+            for (let i = 0; i < chunks.length; i++) {
+                if (chunks[i].number === chunkNum) {
+                    if (status === this.STATUS_RETRY && chunks[i].retry > this.MAX_RETRY) {
+                        chunks[i].status = this.STATUS_FAILED;
+                        ProjectStore.uploadError(uploadId);
+                        return;
+                    }
+                    if (status === this.STATUS_RETRY) chunks[i].retry++;
+                    chunks[i].status = status;
+                    break;
+                }
+            }
+        }
+
+        for (let i = 0; i < chunks.length; i++) {
+            let chunk = chunks[i];
+            if (chunk.status === this.STATUS_WAITING_FOR_UPLOAD || chunk.status === this.STATUS_RETRY) {
+                chunk.status = this.STATUS_UPLOADING;
+                ProjectActions.getChunkUrl(uploadId, upload.blob.slice(chunk.start, chunk.end), chunk.number, upload.size, upload.parentId, upload.parentKind);
```

Comment:
if getChunkUrl fails (ex: dodgy network), this leaves chunk status as uploading, but won't retry and unable to finish. any way to trap around getChunkUrl?